kotlinx-coroutines-core/common/src/CompletionHandler.common.kt

```diff
@@ -1,43 +1,69 @@
 package kotlinx.coroutines
 
-import kotlinx.coroutines.internal.*
-
 /**
  * Handler for [Job.invokeOnCompletion] and [CancellableContinuation.invokeOnCancellation].
  *
- * Installed handler should not throw any exceptions. If it does, they will get caught,
- * wrapped into [CompletionHandlerException], and rethrown, potentially causing crash of unrelated code.
+ * The installed handler should not throw any exceptions.
+ * If it does, they will get caught, wrapped into [CompletionHandlerException], and rethrown, potentially crashing
+ * unrelated code.
  *
- * The meaning of `cause` that is passed to the handler:
- * - Cause is `null` when the job has completed normally.
- * - Cause is an instance of [CancellationException] when the job was cancelled _normally_.
+ * The meaning of `cause` that is passed to the handler is:
+ * - It is `null` if the job has completed normally.
+ * - It is an instance of [CancellationException] if the job was cancelled _normally_.
  *   **It should not be treated as an error**. In particular, it should not be reported to error logs.
  * - Otherwise, the job had _failed_.
  *
+ * A function used for this should not throw any exceptions.
+ * If it does, they will get caught, wrapped into [CompletionHandlerException], and then either
+ * - passed to [handleCoroutineException] for [CancellableContinuation.invokeOnCancellation]
+ *   and, for [Job] instances that are coroutines, [Job.invokeOnCompletion], or
+ * - for [Job] instances that are not coroutines, simply thrown, potentially crashing unrelated code.
+ *
+ * Functions used for this must be fast, non-blocking, and thread-safe.
+ * This handler can be invoked concurrently with the surrounding code.
+ * There is no guarantee on the execution context in which the function is invoked.
+ *
  * **Note**: This type is a part of internal machinery that supports parent-child hierarchies
  * and allows for implementation of suspending functions that wait on the Job's state.
  * This type should not be used in general application code.
- * Implementations of `CompletionHandler` must be fast and _lock-free_.
  */
+// TODO: deprecate. This doesn't seem better than a simple function type.
```

Comment:
Should we?

I see this type alias (or maybe even I would interpret it as an opaque type if it was possible) as a perfect fit -- it is a functional type, but with very specific semantics that is a property of this "type", not of the callsite (`invokeOn*` and coroutine builders), and the documentation to this type is a perfect place to spell it out. 

WDYT? As usual, this particular endpoint might be not the most important one, but it's nice to align our mental models here

Reply:
Here's my reasoning behind the `TODO`.

tl;dr: this function's contract is defined by the places that accept the function, not the other way around, so these other places should define what they expect. Also, the latest commit that fixed the explanation of what the argument means, and this type looks even more unnatural now.

Long version:

I know of two use cases for binding an arbitrary function to a name:
* Passing it somewhere so that that other place calls it,
* Storing it somewhere so that someone calls it later.

For the second use case, introducing a type alias is clearly beneficial: when you have a `var handler: CompletionHandler`, there is *some* contract that should connect the parties. If I see a `CompletionHandler`, I may only invoke it so or so, and return, it promises me this and that, and vice versa. But what we have here is purely the first use case.

> but with very specific semantics that is a property of this "type", not of the callsite

This surprises me greatly: I think it's exactly the opposite.

The contract on the handler is exactly the combination of the contracts of the functions that accept the handler. Here's the deciding factor for me: if the behavior of `invokeOnCompletion` changed and the contract on `CompletionHandler` could somehow be relaxed, we wouldn't define a `RelaxedCompletionHandler` (supertype of `CompletionHandler`), we would just change the documentation of `CompletionHandler`. So, `CompletionHandler` is a "fake" type, a product of circumstances, only useful for passing it to the specific call sites.

Reply:
Thanks for the elaborate answer.

>we would just change the documentation of CompletionHandler. 
>The contract on the handler is exactly the combination of the contracts of the functions that accept the handler.

Indeed. It appears I just think about this one inside out -- its definition/contract is driven bu the need of its "clients" (`invokeOnCompletion`), yet the moment the contract is established -- it's the property of *`CompletionHandler`* that it does not throw, not the vise versa -- "in general, CompletionHandler is whatever you want it to be, but it shouldn't throw if passed to Job.invokeOn...()"

Reply:
It's quite close. I don't see it fit e.g. for an arbitrary function, but here we have a multitude of functions that all accept the same kind of parameter with the same kind of constraints. These constraints are significant enough to document, and broad enough to get tired of repeating it (NB: here, I focus not on "repeating the text" but on "repeating the conceptual complexity of the entity accepting this parameter and draining our complexity budget").

It's just lightweight and specific enough so it doesn't deserve a dedicated nominal type.

(please note that this discussion is not a blocker of this PR by any means)

Reply:
> here we have a multitude of functions that all accept the same kind of parameter with the same kind of constraints

Let's consider two scenarios separately:
* There's a `CompletionHandler` and, separately, a `CancellationHandler`;
* The two types of handlers stay merged into one, just because they have the same type.

I'm confident the second option is a no-go because the constraints and the behavior are actually different between `invokeOnCompletion` and `invokeOnCancellation`. Knowing just the `CancellationHandler` type, you can't make a judgement about how it will behave, you need to know which of the two endpoints ends up actually receiving the handler.

The first option also seems strange to me: then, the contract on `CompletionHandler` would simply completely mirror the contract on `invokeOnCompletion`, and I don't see the benefits of introducing a separate entry point when we can just say "will be passed to `invokeOnCompletion` and must satisfy the same requirements" and in fact, *must* either do so or direct people to the documentation of `CompletionHandler` explicitly, because no one realistically is going to notice, "Aha, this function accepts not just any lambda but one with a special type, let me read the docs."